### DISCUSSION TITLE
having difficulty to parse a test JSON file with this

### DISCUSSION THREAD

**Discussion by creativegamer03** [open-source-parsers/jsoncpp#1323] [Q&A] [answered]
So I was trying to use this parser to recode my Python program to C++ since it had a slower execution time than C++. However, I still can't get it to build and test my code due to errors, all about `does not name a type` errors in most lines from my simple test program.
Here's my code:
```cpp
#include "json/json.h"
#include <fstream>

std::ifstream sample_file("sample.json", std::ifstream::binary);
sample_file >> sample;

// cout << people; //This will print the entire json object.

// The following lines will let you access the indexed objects.
cout << "Bible version: ";
cout << sample["version"]; // bible version
cout << "Sample verse content";
cout << sample["testaments"][0]["books"][2]["chapters"][1]["verses"][1]["verse"]; //sample bible verse selection from JSON file
cout << "Sample footnote content";
cout << sample["testaments"][0]["books"][2]["chapters"][1]["footnotes"][1]["reference"];
cout << sample["testaments"][0]["books"][2]["chapters"][1]["footnotes"][1]["footnotes"]; //sample bible footnote selection from JSON file
```
And here's the console build messages/log:

Line | Message
---- | -------
5 | `error: 'sample_file' does not name a type`
10 | `error: 'cout' does not name a type`
11 | `error: 'cout' does not name a type`
12 | `error: 'cout' does not name a type`
13 | `error: 'cout' does not name a type`
14 | `error: 'cout' does not name a type`
15 | `error: 'cout' does not name a type`
16 | `error: 'cout' does not name a type`

Btw, the code was based from the answer here from stack overflow:
https://stackoverflow.com/questions/32205981/reading-json-files-in-c

And I'm building this from the GNU GCC compiler in Code::Blocks.
Can someone clarify me why this does not work and what's the proper way to do it?

**Comment by BillyDonahue** [comment, marked as answer]
It looks like you skipped over some C++ fundamentals. This forum is not a place to learn C++. I think you need to back up a few steps and get some experience with very basic C++ program structure and then come back to this problem.

All of those `cout` statements need to be `std::cout`, and you need to `#include <iostream>` to see the declaration for `std::cout`.
The bigger problem is that all those statement need to be inside a function. Statements can't appear at file scope like that in C++ like they can in Python, so C++ thinks you're declaring a variable or function, which is why it's complaining that `cout` doesn't name a type. You need to put all those statements into a `main` function and declare the `sample` variable as a `Json::Value sample;` before putting an istream to it.

**Reply by creativegamer03** [reply]
by the time I posted this, I did some modifications on the code since I thought it needs the `main()` function, so I did, including adding some other stuff. By the time I checked your answer, I added the `Json::Value sample;` line.

The code so far:
```cpp
#include "json/json.h"
#include <fstream>
#include <iostream>

int main() {
    Json::Value sample;

    std::ifstream sample_file("sample.json", std::ifstream::binary);
    sample_file >> sample;
    // cout << people; //This will print the entire json object.

    // The following lines will let you access the indexed objects.
    std::cout << "Bible version: ";
    std::cout << sample["version"]; // bible version
    std::cout << "Sample verse content";
    std::cout << sample["testaments"][0]["books"][2]["chapters"][1]["verses"][1]["verse"]; //sample bible verse selection from JSON file
    std::cout << "Sample footnote content";
    std::cout << sample["testaments"][0]["books"][2]["chapters"][1]["footnotes"][1]["reference"];
    std::cout << sample["testaments"][0]["books"][2]["chapters"][1]["footnotes"][1]["footnotes"]; //sample bible footnote selection from JSON file

    return 0;
}
```
Building almost succeeded until it errored on the *.o file it generated after it build an *.exe file but was then removed due to several errors about `undefined reference to 'Json::Value::<several things under this>`. Is it because of the amalgamated version of the jsoncpp package, or is it an error in my code (again)? I used the Amalgamated method to add jsoncpp to my project because the vcpkg method didn't worked for me due to needing a x64 7zip installation, of which my current device doesn't support because it's x86.

**Reply by BillyDonahue** [reply]
You'll have to BUILD the jsoncpp library into a .a file and then link it into your executable. Using libraries in C++ is not so easy. Even with the amalgamated packaging variant of the jsoncpp library, you need to _link_  the jsoncpp library into your executable. The way you do this varies between development environments.

For Code::Blocks, it looks like you need to start here:
https://www.codeblocks.org/docs/main_codeblocks_en3.html#x3-350001.11.13

**Reply by creativegamer03** [reply]
Ok I just successfully built the library into an .a file, and the test json program was finally built. However, it did return a status code 3.
**Console output:**
```
terminate called after throwing an instance of 'Json::RuntimeError'
  what():  * Line 1, Column 1
  Syntax error: value, object or array expected.


Process returned 3 (0x3)   execution time : 9.378 s
Press any key to continue.
```
Is this related to the JSON file, the program's code, or jsoncpp? The json file is good (did a validation check online).
